File: JSPTH/MathJS/trigonometry.js
class Trigonometry {
    sine(x) {
        return Math.sin(x);
    }

    cosine(x) {
        return Math.cos(x);
    }

    tangent(x) {
        return Math.tan(x);
    }

    secant(x) {
        if (Math.cos(x) === 0) throw new Error("Secant undefined at this angle");
        return 1 / Math.cos(x);
    }

    cosecant(x) {
        if (Math.sin(x) === 0) throw new Error("Cosecant undefined at this angle");
        return 1 / Math.sin(x);
    }

    cotangent(x) {
        if (Math.tan(x) === 0) throw new Error("Cotangent undefined at this angle");
        return 1 / Math.tan(x);
    }
    arcsine(x) {
        return Math.asin(x);
    }

    arccosine(x) {
        return Math.acos(x);
    }

    arctangent(x) {
        return Math.atan(x);
    }

    law_of_cosines(a, b, angleC) {
        return Math.sqrt(a * a + b * b - 2 * a * b * Math.cos(angleC));
    }

    law_of_sines(a, angleA, angleB) {
        return (a * Math.sin(angleB)) / Math.sin(angleA);
    }

    sine_sum(x, y) {
        return Math.sin(x) * Math.cos(y) + Math.cos(x) * Math.sin(y);
    }

    cosine_sum(x, y) {
        return Math.cos(x) * Math.cos(y) - Math.sin(x) * Math.sin(y);
    }

    tangent_sum(x, y) {
        if (Math.cos(x) * Math.cos(y) === 0) throw new Error("Tangent sum undefined for these angles");
        return (Math.tan(x) + Math.tan(y)) / (1 - Math.tan(x) * Math.tan(y));
    }


    sine_double(x) {
        return 2 * Math.sin(x) * Math.cos(x);
    }

    cosine_double(x) {
        return Math.cos(x) * Math.cos(x) - Math.sin(x) * Math.sin(x);
    }

    tangent_double(x) {
        if (Math.cos(x) * Math.cos(x) === 0) throw new Error("Tangent double undefined for this angle");
        return (2 * Math.tan(x)) / (1 - Math.tan(x) * Math.tan(x));
    }

    polar_to_cartesian(radius, angle) {
        return {
            x: radius * Math.cos(angle),
            y: radius * Math.sin(angle),
        };
    }

    cartesian_to_polar(x, y) {
        return {
            radius: Math.sqrt(x * x + y * y),
            angle: Math.atan2(y, x),
        };
    }
}

File: Rust/JSRS/src/JSRS/Math/Trigonometry.rs
// src/main.rs

use std::f64::consts::PI;

struct TrigonometrySolver;

impl TrigonometrySolver {
    /// Converts degrees to radians.
    fn to_radians(degrees: f64) -> f64 {
        degrees * PI / 180.0
    }

    /// Converts radians to degrees.
    fn to_degrees(radians: f64) -> f64 {
        radians * 180.0 / PI
    }

    /// Basic trigonometric functions
    fn sine(angle: f64) -> f64 {
        angle.sin()
    }

    fn cosine(angle: f64) -> f64 {
        angle.cos()
    }

    fn tangent(angle: f64) -> f64 {
        angle.tan()
    }

    fn secant(angle: f64) -> f64 {
        1.0 / angle.cos()
    }

    fn cosecant(angle: f64) -> f64 {
        1.0 / angle.sin()
    }

    fn cotangent(angle: f64) -> f64 {
        1.0 / angle.tan()
    }

    /// Inverse trigonometric functions
    fn arcsine(value: f64) -> f64 {
        value.asin()
    }

    fn arccosine(value: f64) -> f64 {
        value.acos()
    }

    fn arctangent(value: f64) -> f64 {
        value.atan()
    }

    /// Law of Cosines: c^2 = a^2 + b^2 - 2ab * cos(C)
    fn law_of_cosines(a: f64, b: f64, angle_c: f64) -> f64 {
        (a.powi(2) + b.powi(2) - 2.0 * a * b * angle_c.cos()).sqrt()
    }

    /// Law of Sines: a / sin(A) = b / sin(B) = c / sin(C)
    fn law_of_sines(a: f64, angle_a: f64, angle_b: f64) -> f64 {
        a * angle_b.sin() / angle_a.sin()
    }

    /// Angle sum and double angle formulas
    fn sine_sum(angle1: f64, angle2: f64) -> f64 {
        angle1.sin() * angle2.cos() + angle1.cos() * angle2.sin()
    }

    fn cosine_sum(angle1: f64, angle2: f64) -> f64 {
        angle1.cos() * angle2.cos() - angle1.sin() * angle2.sin()
    }

    fn tangent_sum(angle1: f64, angle2: f64) -> f64 {
        (angle1.tan() + angle2.tan()) / (1.0 - angle1.tan() * angle2.tan())
    }

    fn sine_double(angle: f64) -> f64 {
        2.0 * angle.sin() * angle.cos()
    }

    fn cosine_double(angle: f64) -> f64 {
        1.0 - 2.0 * angle.sin().powi(2)
    }

    fn tangent_double(angle: f64) -> f64 {
        2.0 * angle.tan() / (1.0 - angle.tan().powi(2))
    }

    /// Polar to Cartesian coordinates conversion
    fn polar_to_cartesian(radius: f64, angle: f64) -> (f64, f64) {
        let x = radius * angle.cos();
        let y = radius * angle.sin();
        (x, y)
    }

    /// Cartesian to Polar coordinates conversion
    fn cartesian_to_polar(x: f64, y: f64) -> (f64, f64) {
        let radius = (x.powi(2) + y.powi(2)).sqrt();
        let angle = y.atan2(x);
        (radius, angle)
    }
}
